doc/release-notes/release-6.sgml

```diff
@@ -63,7 +63,9 @@ This section gives an account of those changes in three categories:
 <sect1>Changes to existing directives<label id="modifieddirectives">
 <p>
 <descrip>
-	<p>There have been no directives changed.
+	<tag>ssl_engine</tag>
+	<p>OpenSSL 3.0.0 removes the Engine feature. This directive is
+	   only supported when Squid is built for older OpenSSL versions.
 
```

Comment:
N.B. This is a copy the last [comment](https://github.com/squid-cache/squid/pull/694#discussion_r769077272
) from an earlier change request that was silently "resolved" for some (unknown to me) reason. I crossed out item 3 below because that change has been done, but we still have not even established whether engine support should be deprecated/disabled (a prerequisite for those to-do items). I will now re-resolve that outdated change request so that we have only one conversation to worry about.

> The ENGINE API is not available in the libssl3 I am building/testing against.

The limitations specific to the library you use should not affect the decision regarding deprecation of a Squid feature that may be used by others. Perhaps you can rephrase the above statement in more general terms (e.g., OpenSSL v3 that comes by default with Ubuntu X disables ENGINE API support)? A _general_ statement about the necessary library availability could influence our decision to deprecate, of course.

> I have never seen this option used by anyone. So for now I have opted to deprecate the whole feature.

You have seen it used by bug [4112](https://bugs.squid-cache.org/show_bug.cgi?id=4112) reporter (at least) AFAICT.

> If anyone really wants to have an ENGINE with libssl3 they will need to refactor all the code making use of this config option and the variables it sets up.

The cost of support is a valid factor in our decision making process, of course. Can you give an example of Squid code that will need to be changed in order to preserve support for OpenSSL engine selection via `ssl_engine` directive in Squid built with OpenSSL v3? My superficial interpretation of OpenSSL v3 migration [documentation](https://www.openssl.org/docs/man3.0/man7/migration_guide.html) is that there should be no Squid changes as long as the OpenSSL library is built with that deprecated API/code enabled.

----

Assuming the continued support is indeed costly (which is to be determined above), one could address this change request by making the following changes:

1. Adjust PR description to document that we are deprecating OpenSSL Engine support (not because it is removed in OpenSSL v3 but) because it is deprecated in OpenSSL v3, popular distributions ship OpenSSL v3 without Engine support (if that is true), and we do not have the resources to update the corresponding Squid code to remain compatible with OpenSSL v3 (if the code needs updating).
2. Add deprecation warning (for OpenSSL before v3) to squid.conf parser or interpreter. You have already adjusted the code to fail with OpenSSL v3+.
3. ~Replace the current (incorrect) claim in cf.data.pre with something along these lines: "Not supported in builds with OpenSSL v3 or newer."~

There are other ways to address this change request, of course. The above is just an example that illustrates the missing pieces if we decide to deprecate `ssl_engine` support starting with OpenSSL v3 builds.

Reply:
> N.B. This is a copy the last [comment](https://github.com/squid-cache/squid/pull/694#discussion_r879784328) from an earlier [change request](https://github.com/squid-cache/squid/pull/694#discussion_r769077272) that was silently "resolved" for some (unknown to me) reason. 

github says that thread was:
> rousskov marked this conversation as resolved.

Reply:
Do you find this working clearer?
```
 OpenSSL 3.0.0 replaces the Engine feature Providers (not yet supported by Squid).
 This directive is only supported when Squid is built for older OpenSSL versions.
```

Reply:
> github says that "rousskov marked this conversation as resolved"

Yes, IIRC, I unresolved the old (already incorrectly resolved at that time) thread, but decided to move the conversation here instead, and re-resolved the old thread after the move. What you are seeing is the latest state. Unfortunately, these GitHub pages do not reflect the full resolution history (AFAICT), just its outcome.

> Do you find this working clearer?

As far as clarity is concerned, a word "with" appears to be missing after "feature" in that proposed text, but, more importantly, I do not understand your resistance to telling folks the truth about OpenSSL deprecation of the Engine feature. The feature is still available in OpenSSL v3. It is deprecated. We decided to drop its support now because, in part, keeping that support required significant[^1] work. There is nothing wrong with any of that if it is all true! We do not need to hide any of those (potentially important to some Squid admins) facts behind vague phrasing. We can and should be open/direct here IMO.

[^1]: We have not yet established that the amount of work to keep Engines support is significant, but I assume, for the sake of the argument, that we will establish that.

Reply:
> I do not understand your resistance to telling folks the truth about OpenSSL deprecation of the Engine feature. The feature is still available in OpenSSL v3. It is deprecated.

The API used by Squid to load an ENGINE is not deprecated. It is **gone**.

AFAICT, what has been deprecated but still works are the APIs for ENGINES to pass data/object pointers to-from libssl. **after** being refactored to load through the Provider API old ENGINE's can continue to use the deprecated APIs to exchange data.  None of that has anything to do with Squid.

None of these OpenSSL-internal technicalities are of any consequence to Squid admin.

Reply:
Would it make sense to deprecate the whole config option instead of dealing with the complexity?
If it's an useless internal technicality, we could ignore-and-warn

Reply:
> Would it make sense to deprecate the whole config option instead of dealing with the complexity?

Yes, that is item 2 on the list in this change request (after the complexity claim has been substantiated).

The cost of satisfying the preconditions for that deprecation has been artificially inflated by lack of cooperation, but we are making progress and will get there eventually.

>  If it's an useless internal technicality, we could ignore-and-warn

Justification of a (known to be used in the past) feature deprecation is not a useless technicality. We ought to be clear about such a deprecation (internally and externally). I have been [asking](https://github.com/squid-cache/squid/pull/694#discussion_r769077272) for that clarity since December 2021, but, again, we are making progress and will eventually get there.

Reply:
As stated already several times there is no OpenSSL API replacement to fix these:
```
../../../src/ssl/support.cc: In function 'void Ssl::Initialize()':
../../../src/ssl/support.cc:662:36: error: 'void ENGINE_load_builtin_engines()' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  662 |         ENGINE_load_builtin_engines();
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from ../../../src/ssl/support.h:30,
                 from ../../../src/AccessLogEntry.h:32,
                 from ../../../src/acl/FilledChecklist.h:12,
                 from ../../../src/ssl/support.cc:18:
/usr/include/openssl/engine.h:358:28: note: declared here
  358 | OSSL_DEPRECATEDIN_3_0 void ENGINE_load_builtin_engines(void);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/ssl/support.cc:664:31: error: 'ENGINE* ENGINE_by_id(const char*)' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  664 |         if (!(e = ENGINE_by_id(::Config.SSL.ssl_engine)))
      |                   ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../src/ssl/support.h:30,
                 from ../../../src/AccessLogEntry.h:32,
                 from ../../../src/acl/FilledChecklist.h:12,
                 from ../../../src/ssl/support.cc:18:
/usr/include/openssl/engine.h:336:31: note: declared here
  336 | OSSL_DEPRECATEDIN_3_0 ENGINE *ENGINE_by_id(const char *id);
      |                               ^~~~~~~~~~~~
../../../src/ssl/support.cc:667:32: error: 'int ENGINE_set_default(ENGINE*, unsigned int)' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  667 |         if (!ENGINE_set_default(e, ENGINE_METHOD_ALL)) {
      |              ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../src/ssl/support.h:30,
                 from ../../../src/AccessLogEntry.h:32,
                 from ../../../src/acl/FilledChecklist.h:12,
                 from ../../../src/ssl/support.cc:18:
/usr/include/openssl/engine.h:708:27: note: declared here
  708 | OSSL_DEPRECATEDIN_3_0 int ENGINE_set_default(ENGINE *e, unsigned int flags);
      |                           ^~~~~~~~~~~~~~~~~~
```

Reply:
I have now also reviewed the Providers API to see how it differs from ENGINE.

The conclusion I have come to is that we have two options:

 1)  in squid.conf add a new ```providers=`` option to each of the directives which create a TLS/SSL context (*_port, cache_peer, tls_outgoing_options) . Since the set of Providers loaded into an interface specific context determines what ```tls-options=``` are available and how they work on TLS connections using that context (eg ciphers available).

2) leave Provider support out of Squid entirely and encourage admin use openssl.conf to configure them. This IMO is better overall since a) engines and providers are mostly useful to add a hardware device or enable FIPS mode at the library level, and b) it reduces our own work and code supporting the feature.

Reply:
I would favour option 2, because it's simpler to implement and maintain, and it is more future-proof, as it will leave openssl in charge of configuring their own settings. At most we should reference this in our documentation

Reply:
In effort to make some progress a recap:

Alex:
> what would it take to preserve legacy Engines support while using OpenSSL v3.0?

Alex:
> Can you give an example of Squid code that will need to be changed in order to preserve support for OpenSSL engine selection via ssl_engine directive in Squid built with OpenSSL v3?


Amos:
> there is no OpenSSL API replacement to fix these:
> `OSSL_DEPRECATEDIN_3_0 void ENGINE_load_builtin_engines()`
> `OSSL_DEPRECATEDIN_3_0 ENGINE *ENGINE_by_id()`
> `OSSL_DEPRECATEDIN_3_0 int ENGINE_set_default()`
> 

Alex:
> I am not sure why you are stating the above. Your statement does not seem to invalidate or address this change request.

You elided the important part of what I quoted ... The fact that those symbols use produce fatal build errors.